packages/next-swc/crates/next-core/src/next_app/mod.rs

```diff
@@ -6,12 +6,283 @@ pub(crate) mod app_page_entry;
+)]
+pub struct AppPage(pub Vec<PageSegment>);
+
+impl AppPage {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn push(&mut self, segment: PageSegment) -> Result<()> {
+        if matches!(
+            self.0.last(),
+            Some(PageSegment::CatchAll(..) | PageSegment::OptionalCatchAll(..))
+        ) && !matches!(segment, PageSegment::PageType(..))
+        {
+            bail!(
+                "Invalid segment {}, catch all segment must be the last segment",
+                segment
+            )
+        }
+
+        self.0.push(segment);
+        Ok(())
+    }
+
+    pub fn push_str(&mut self, segment: &str) -> Result<()> {
+        if segment.is_empty() {
+            return Ok(());
+        }
+
+        self.push(PageSegment::parse(segment)?)
+    }
+
+    pub fn clone_push(&self, segment: PageSegment) -> Result<Self> {
+        let mut cloned = self.clone();
+        cloned.push(segment)?;
+        Ok(cloned)
+    }
+
+    pub fn clone_push_str(&self, segment: &str) -> Result<Self> {
+        let mut cloned = self.clone();
+        cloned.push_str(segment)?;
+        Ok(cloned)
+    }
+
+    pub fn parse(page: &str) -> Result<Self> {
+        let mut app_page = Self::new();
+
+        for segment in page.split('/') {
+            app_page.push_str(segment)?;
+        }
+
+        Ok(app_page)
+    }
+}
+
+impl Display for AppPage {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        if self.0.is_empty() {
+            return f.write_char('/');
+        }
+
+        for segment in &self.0 {
+            f.write_char('/')?;
+            Display::fmt(segment, f)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Deref for AppPage {
+    type Target = [PageSegment];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+#[derive(Clone, Debug, Hash, Serialize, Deserialize, PartialEq, Eq, TaskInput, TraceRawVcs)]
+pub enum PathSegment {
+    Static(String),
+    Dynamic(String),
+    CatchAll(String),
+    OptionalCatchAll(String),
+}
+
+impl Display for PathSegment {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PathSegment::Static(s) => f.write_str(s),
+            PathSegment::Dynamic(s) => {
+                f.write_char('[')?;
+                f.write_str(s)?;
+                f.write_char(']')
+            }
+            PathSegment::CatchAll(s) => {
+                f.write_str("[...")?;
+                f.write_str(s)?;
+                f.write_char(']')
+            }
+            PathSegment::OptionalCatchAll(s) => {
+                f.write_str("[[...")?;
+                f.write_str(s)?;
+                f.write_str("]]")
+            }
+        }
+    }
+}
+
+/// The pathname (including dynamic placeholders) for a route to resolve.
+#[derive(
+    Clone, Debug, Hash, PartialEq, Eq, Default, Serialize, Deserialize, TaskInput, TraceRawVcs,
+)]
+pub struct AppPath(pub Vec<PathSegment>);
```

Comment:
Is the difference just that routes have a more restricted path meta syntax?

Reply:
`AppPath` is the path used by the router, `AppPage` includes groups and parallel routes (basically mirrors the file system)